server/src/main/java/org/candlepin/controller/ActualRefresher.java

```diff
@@ -0,0 +1,1297 @@
+        3       non-null            null                true                        reference update (affected entity)
+        4       non-null            null                false                       skip
+        5       null                non-null            true                        entity creation
+        6       null                non-null            false                       entity creation
+        7       null                null                true                        - ERROR -
+        8       null                null                false                       - ERROR -
+
+        */
+
+        // Check if we have changes
+        if (node.changed()) {
+            if (!node.hasExistingEntity()) {
+                // Creation (5, 6)
+
+                switch (node.getType()) {
+                    case PRODUCT:
+                        node.setUpdatedEntity(this.createEntity(mapper, vMap,
+                            (Product) null,
+                            (Product) node.getImportedEntity()));
+
+                        break;
+
+                    case CONTENT:
+                        node.setUpdatedEntity(this.createEntity(mapper, vMap,
+                            (Content) null,
+                            (Content) node.getImportedEntity()));
+
+                        break;
+
+                    default:
+                        throw new IllegalStateException("Unexpected node type: " + node.getType());
+                }
+            }
+            else {
+                // Update (1, 2)
+
+                switch (node.getType()) {
+                    case PRODUCT:
+                        node.setUpdatedEntity(this.createEntity(mapper, vMap,
+                            (Product) node.getExistingEntity(), (ProductInfo) node.getImportedEntity()));
+
+                        break;
+
+                    case CONTENT:
+                        node.setUpdatedEntity(this.createEntity(mapper, vMap,
+                            (Content) node.getExistingEntity(), (ContentInfo) node.getImportedEntity()));
+
+                        break;
+
+                    default:
+                        throw new IllegalStateException("Unexpected node type: " + node.getType());
+                }
+            }
+        }
+        else if (childrenUpdated) {
+            node.markChanged(); // probably unnecessary, but might as well be consistent
+
+            // Resolve children references
+            switch (node.getType()) {
+                case PRODUCT:
+                    node.setUpdatedEntity(this.createEntity(mapper, vMap,
+                        (Product) null,
+                        (Product) node.getImportedEntity()));
+
+                    break;
+
+                case CONTENT:
+                    node.setUpdatedEntity(this.createEntity(mapper, vMap,
+                        (Content) null,
+                        (Content) node.getImportedEntity()));
```

Comment:
At this point, we're in scenario 3, so the importedEntity is null, while the existingEntity is not null; shouldn't this be like this then?:
 ```
                    node.setUpdatedEntity(this.createEntity(mapper, vMap,
                        (Content) node.getExistingEntity(),
                        (Content) null));
```

Reply:
Yeah, that sounds right.